torch/distributed/fsdp/_symbolic_trace.py

```diff
@@ -0,0 +1,241 @@
+    concrete_args: Optional[Dict[str, Any]] = None
+
+
+@dataclass
+class _ExecutionInfo:
+    """
+    Contains the execution order information in the model forward pass.
+
+    Attributes:
+        current_module: record the module that is currently being traced.
+
+        module_forward_order: a list of modules, where the ordering is based on
+            when their forward function is called.
+
+        param_exec_order: a list of parameters ordered based on their execution order.
+
+        _traced_param_set: a set containing all parameters that have been traced.
+
+        module_to_execution_infos: a dict that maps each module to a list of
+            tuples each containing a module and a list of named parameters.
+            For a given module, each tuple:
+            1. either contains this module and part of its ``named_parameters`` that will be executed together,
+            2. or contains one of its child modules and all of the child module's ``named_parameters``.
+            The list of tuples is ordered based on the parameter execution order.
+    """
+
+    current_module: torch.nn.Module
+    module_forward_order: List[torch.nn.Module]
+    module_to_execution_infos: Dict[
+        torch.nn.Module,
+        List[Tuple[torch.nn.Module, List[Tuple[str, torch.nn.Parameter]]]],
+    ]
+    param_exec_order: List[torch.nn.Parameter] = field(default_factory=list)
+    _traced_param_set: Set[torch.nn.Parameter] = field(default_factory=set)
+
+
+def _init_execution_info(root_module: torch.nn.Module) -> _ExecutionInfo:
+    """
+    Create an instance of _ExecutionInfo with initialization based on ``root_module``.
+
+    Args:
+        root_module (torch.nn.Module): the module to get the execution information
+            via ``tracer.trace()`` inside ``_patch_tracer``.
+    """
+    return _ExecutionInfo(
+        current_module=root_module,
+        module_forward_order=[root_module],
+        module_to_execution_infos={root_module: []},
+    )
+
+
+def _patched_create_proxy(
+    create_proxy: Callable,
+    execution_info: _ExecutionInfo,
+    params_dict: Dict[str, torch.nn.Parameter],
+    kind: str,
+    target: torch.fx.node.Target,
+    args: Tuple[Any, ...],
+    kwargs: Dict[str, Any],
+    name: Optional[str] = None,
+    type_expr: Optional[Any] = None,
+    proxy_factory_fn: Callable[[torch.fx.Node], torch.fx.Proxy] = None,
+) -> torch.fx.Proxy:
+    """
+    Override of :meth:`~torch.fx.Tracer.create_proxy`.
+    ``Tracer.create_proxy`` is called in symbolic tracing for each leaf function/method/module.
+    This override intercepts the recording of each of these operations to
+    update ``execution_info.module_to_execution_infos``.
+
+    Args:
+        create_proxy (Callable):
+            The ``create_proxy`` function to be patched.
+        execution_info (_ExecutionInfo):
+            Used to record the execution information.
+        params_dict (Dict[str, torch.nn.Parameter]):
```

Comment:
nit: Kind of like the other PR in my branch, from reading the name `params_dict`, I am not sure what are the keys and values and would have to trace the code to this docstring. Personally, I would name this `prefixed_param_name_to_param`. (I specifically would say `prefixed_param_name` since sometimes we also track the non-prefixed parameter name.)